### PR TITLE
EC-301 blockchain recovery from levelDB

### PR DESCRIPTION
**Pull request by GRYE**

### CODE REVIEW

File: src/main/scala/encry/EncryApp.scala [hunk 1]
```diff
@@ -52,7 +52,7 @@ object EncryApp extends App with Logging {
   lazy val upnp: UPnP = new UPnP(settings.network)
 
   lazy val nodeViewHolder: ActorRef = system.actorOf(EncryNodeViewHolder.props()
-    .withDispatcher("nvh-dispatcher").withMailbox("nvh-mailbox"), "nodeViewHolder")
+    .withDispatcher("nvh-dispatcher"), "nodeViewHolder")
```
Comment (kapinuss) [inline src/main/scala/encry/EncryApp.scala]: return mailbox

File: src/main/scala/encry/network/ModifiersHolder.scala [hunk 2]
```diff
@@ -20,18 +20,25 @@ class ModifiersHolder extends PersistentActor with Logging {
 
   /** Map, which contains not completed blocks
     * Key can be payloadId or also header id. So value depends on key, and can contains: headerId, payloadId */
-  var headers: Map[String, (EncryBlockHeader, Int)] = Map.empty
-  var payloads: Map[String, (EncryBlockPayload, Int)] = Map.empty
-  var nonCompletedBlocks: Map[String, String] = Map.empty
-  var completedBlocks: SortedMap[Int, EncryBlock] = SortedMap.empty
+  private var headers: Map[String, (EncryBlockHeader, Int)] = Map.empty
```
Comment (kapinuss) [inline src/main/scala/encry/network/ModifiersHolder.scala]: don't use private on actors please

File: src/main/scala/encry/network/ModifiersHolder.scala [hunk 3]
```diff
@@ -20,18 +20,25 @@ class ModifiersHolder extends PersistentActor with Logging {
...
+  private var payloads: Map[String, (EncryBlockPayload, Int)] = Map.empty
+  private var nonCompletedBlocks: Map[String, String] = Map.empty
+  private var completedBlocks: SortedMap[Int, EncryBlock] = SortedMap.empty
+  private val batchSize: Int = 400
```
Comment (kapinuss) [inline src/main/scala/encry/network/ModifiersHolder.scala]: to the config

File: src/main/scala/encry/network/peer/PeerManager.scala [hunk 4]
```diff
@@ -16,8 +17,9 @@ import scala.util.Random
 
 class PeerManager extends Actor with Logging {
 
-  var connectedPeers: Map[InetSocketAddress, ConnectedPeer] = Map.empty
-  var connectingPeers: Set[InetSocketAddress] = Set.empty
+  private var connectedPeers: Map[InetSocketAddress, ConnectedPeer] = Map.empty
```
Comment (kapinuss) [inline src/main/scala/encry/network/peer/PeerManager.scala]: - private

File: src/main/scala/encry/view/EncryNodeViewHolder.scala [hunk 5]
```diff
@@ -108,6 +106,15 @@ class EncryNodeViewHolder[StateType <: EncryState[StateType]] extends Actor with
     case a: Any => logError("Strange input: " + a)
   }
 
+  private def computeApplications(): Unit = {
```
Comment (kapinuss) [inline src/main/scala/encry/view/EncryNodeViewHolder.scala]: parenthesis

File: src/main/scala/encry/view/ModifiersCache.scala [hunk 6]
```diff
@@ -144,6 +144,13 @@ class DefaultModifiersCache[PMOD <: EncryPersistentModifier, HR <: EncryHistoryR
 case class EncryModifiersCache(override val maxSize: Int)
   extends DefaultModifiersCache[EncryPersistentModifier, EncryHistory](maxSize) {
 
+  override def put(key: K, value: V): Unit = {
```
Comment (kapinuss) [inline src/main/scala/encry/view/ModifiersCache.scala]: parenthesis

File: src/main/scala/encry/view/history/processors/BlockHeaderProcessor.scala [hunk 7]
```diff
@@ -21,7 +21,7 @@ import scala.annotation.tailrec
 import scala.collection.immutable
 import scala.util.Try
 
-trait BlockHeaderProcessor extends Logging {
+trait BlockHeaderProcessor extends Logging { //scalastyle:ignore
```
Comment (kapinuss) [inline src/main/scala/encry/view/history/processors/BlockHeaderProcessor.scala]: delete comment

File: src/main/scala/encry/view/history/processors/BlockHeaderProcessor.scala [hunk 8]
```diff
@@ -117,44 +117,51 @@ trait BlockHeaderProcessor extends Logging {
...
-    bestHeaderIdOpt match {
-      case Some(bestHeaderId) =>
-        // If we verify transactions, we don't need to send this header to state.
-        // If we don't and this is the best header, we should send this header to state to update state root hash
-        val toProcess: Seq[EncryBlockHeader] = if (nodeSettings.verifyTransactions || !(bestHeaderId sameElements h.id)) Seq.empty else Seq(h)
-        ProgressInfo(None, Seq.empty, toProcess, toDownload(h))
-      case None =>
-        logError("Should always have best header after header application")
-        EncryApp.forceStopApplication()
+    val dataToUpdateOpt: Option[(Seq[(ByteArrayWrapper, ByteArrayWrapper)], EncryPersistentModifier)] = getHeaderInfoUpdate(h)
```
Comment (kapinuss) [inline src/main/scala/encry/view/history/processors/BlockHeaderProcessor.scala]: use /n in all places

File: src/main/scala/encry/view/history/processors/BlockHeaderProcessor.scala [hunk 9]
```diff
@@ -116,45 +116,46 @@ trait BlockHeaderProcessor extends Logging {
...
-        if (score > bestHeadersChainScore) Seq(BestHeaderKey -> ByteArrayWrapper(h.id)) else Seq.empty
-      val scoreRow: (ByteArrayWrapper, ByteArrayWrapper) = headerScoreKey(h.id) -> ByteArrayWrapper(score.toByteArray)
-      val heightRow: (ByteArrayWrapper, ByteArrayWrapper) = headerHeightKey(h.id) -> ByteArrayWrapper(Ints.toByteArray(h.height))
-      val headerIdsRow: Seq[(ByteArrayWrapper, ByteArrayWrapper)] = if (score > bestHeadersChainScore) {
-        bestBlockHeaderIdsRow(h, score)
-      } else {
-        EncryApp.system.actorSelection("/user/blockListener") ! NewOrphaned(h) // TODO: Remove direct system import when possible.
-        orphanedBlockHeaderIdsRow(h, score)
+      scoreOf(h.parentId).map { parentScore =>
+        val score = Difficulty @@ (parentScore + difficulty)
```
Comment (oskin1) [inline src/main/scala/encry/view/history/processors/BlockHeaderProcessor.scala]: Type annotation

**Pull request merged**